### ISSUE TITLE
Quarto Visual Editor support for reference links

### ISSUE THREAD

**Issue by reagle** [quarto-dev/quarto#42]
I'm eagerly awaiting Quarto's stand-alone markdown editor (is there a timeframe on that?) because in my quick test, RStudio's visual mode, is an excellent markdown editor: it supports multiline tables and div sections! Aside from not needing all the R stuff, the show-stopper, though, is that it [rewrites all of my carefully made reference links](https://rstudio.github.io/visual-markdown-editing/markdown.html#known-limitations). Could these be supported?

**Comment by jjallaire** [comment]
We currently aren't able to support reference links because when Pandoc parses the page they are all resolved into "real" links (i.e. the Pandoc AST doesn't have the concept of a reference link, and since our editor is really editing the AST not the markdown text, there isn't any way for us to read or write them).

I'm going to close this issue for now b/c there isn't a way for us to resolve it. Rest assured though if at some point we do get Pandoc AST support for reference links we will implement this in the visual editor!

We do now have a version of the Quarto Visual Editor available in our VS Code extension. Details are here: https://quarto.org/docs/visual-editor/vscode/

**Comment by reagle** [comment]
Hi @jjallaire, okay, I'll investigate/raise this issue with pandoc. Did I misunderstand that there will be a stand alone editor or is there a VSC extension only?

**Comment by jjallaire** [comment]
There will also be a stand alone editor (sometime later this year)

**Comment by reagle** [comment]
Do you not support reference links at all? There's a couple of variations and @jgm confirmed pandoc AST doesn't preserve the label, but perhaps you could rewrite with option 1 or 3?

1. [reference link] shortcut version
2. [reference link][label-a] with label
3. [reference link][] with implicit label

[reference link]: https://example.com/
[label-a]: https://example.com/a

**Comment by jjallaire** [comment]
Pandoc itself is doing the writing (we don't do raw text transformations ourselves). The problem is that we don't have a way of saying to Pandoc "please write this as a reference link" (at least that I know of). Even if we did they wouldn't round-trip back into the AST.

Very open to any path you can lead us to get this working (definitely could be missing something).

**Comment by reagle** [comment]
That option is available on the command line and implicit reference links seem to survive round trips

    pandoc --from markdown --to markdown --wrap=preserve --reference-links oc-syllabus-SP.md > out.md

For example:

https://pandoc.org/try/?params={%22text%22:%22This+is+a+%5Btest%5D.+%5Btest%5D%3A+https%3A%2F%2Fexample.com%2Ftest%22,%22to%22:%22markdown%22,%22from%22:%22markdown%22,%22standalone%22:false,%22embed-resources%22:false,%22table-of-contents%22:false,%22number-sections%22:false,%22citeproc%22:false,%22html-math-method%22:%22plain%22,%22wrap%22:%22auto%22,%22highlight-style%22:null,%22files%22:{},%22template%22:null,%22reference-links%22:true}

**Comment by jgm** [comment]
To clarify: pandoc will allow you to render ALL the links as reference links. But you can't pick and choose.

**Comment by reagle** [comment]
Even if all links are made reference links, I think that makes for much more (human) readable and editable markdown.

**Comment by jjallaire** [comment]
I just test drove this locally and sure enough it does round-trip! We could provide an option similarly to what we currently do for where to write footnotes and how to handle line breaks. I'll re-open this and we'll hopefully work on this very soon!

**Comment by reagle** [comment]
@jjallaire, great, thank you. FYI: [here's the markdown, a syllabus,](https://reagle.org/joseph/2023/cda/cda-syllabus-SP.md) that I tested your editor on. Perhaps it, or other examples I'm happy to share, would be good tests for ensuring whatever you save remains as human legible/editable as possible.

For example, one other nuisance when I go from md2md with pandoc is HTML comments become quoted and labeled. I haven't figured out how to disable that yet, but it means to uncomment something I have to manually select and edit the text rather than using the "uncomment" keyboard shortcut in my editor.

```{=html}
<!-- TODO: this day should be one week after persuasion assignment -->
```

**Comment by reagle** [comment]
Thanks, that's:

```pandoc --from markdown --to markdown-raw_attribute --wrap=preserve --reference-links```

**Comment by jjallaire** [comment]
Support for writing reference links has been added here: https://github.com/quarto-dev/quarto/commit/8db11d6f717a73500dd4d749bee96a1b4e5627e8

This is available in v1.64.0 of the VS Code extension (docs here: https://quarto.org/docs/visual-editor/vscode/#links)
